Add span profiles documentation (#2982)
grafana/pyroscope: docs/sources/configure-client/go-span-profiles.md

---

> title: Go span profiles with trace-to-profiles
> sidebar_label: Go span profiles
> description: Go span profiles with trace-to-profiles
> weight: 100
> ---
> 
> # Go span profiles with trace-to-profiles

Reviewer: is this for the Go language? Otherwise, it can be ready as "Go (for) span profiles with traces to profiles" 

Would it be clearer to use something like Span profiles for Go? Also, Traces to profiles isn't hyphenated in our documentation nor in the UI. 

What do you think about this? Please feel free to edit as you see fit

```suggestion
title: Span profiles with Traces to profiles for Go
menuTitle: Span profiles with Traces to profiles (Go)
description: Learn about and configure Span profiles with Trace to profiles in Grafana for the Go language.
weight: 100
---

# Span profiles with Traces to profiles for Go
```

Author: yeah this is just for go

![span-profiles screenshot](https://grafana.com/static/img/docs/tempo/profiles/tempo-profiles-Span-link-profile-data-source.png)

## Prerequisites

> 1. Profiling: You will need to have your application instrumented with Pyroscopes Go SDK. If you haven't done this yet, please refer to the [Go (push mode)]({{< relref "../configure-client/language-sdks/go_push" >}}) guide.

Reviewer 2: Just a note: strictly speaking, use of SDK is not required as tracing integration also works in pull mode. However, I do support the idea that we tell users to use SDK, if they want to integrate with tracing, because in this way it's just easier to control which Pyroscope labels match trace tags, e.g., using the same `service_name`, etc

> ## Prerequisites
> 
> 1. Profiling: You will need to have your application instrumented with Pyroscopes Go SDK. If you haven't done this yet, please refer to the [Go (push mode)]({{< relref "../configure-client/language-sdks/go_push" >}}) guide.
> 2. Tracing: You will need to have your application instrumented with OpenTelemetry traces. If you haven't done this yet, please refer to the [OpenTelemetry](https://opentelemetry.io/docs/go/getting-started/) guide.

Reviewer: Instead of Prerequisites, we use Before you begin. It's an equivalent title. 

```suggestion
## Before you begin

You need to have profiling and tracing set up before you can use span profiles. 

* Profiling: Your application must be instrumented with Pyroscopes Go SDK. If you haven't done this yet, please refer to the [Go (push mode)]({{< relref "../configure-client/language-sdks/go_push" >}}) guide.
* Tracing: Your application must be instrumented with OpenTelemetry traces. If you haven't done this yet, please refer to the [OpenTelemetry](https://opentelemetry.io/docs/go/getting-started/) guide.
```

This package is a `TracerProvider` implementation, that labels profiling data with span IDs which makes it possible to query for span-specific profiling data in Grafana Tempo UI. 

```shell
# Make sure you also upgrade pyroscope server to version 0.14.0 or higher.
go get github.com/grafana/otel-profiling-go
```

Next you need to create and configure the tracer provider:
```go
package main

import (
	otelpyroscope "github.com/grafana/otel-profiling-go"
	"github.com/grafana/pyroscope-go"
)

func main() {
	// Initialize your tracer provider as usual.
	tp := initTracer()

	// Wrap it with otelpyroscope tracer provider.
	otel.SetTracerProvider(otelpyroscope.NewTracerProvider(tp))

	// If you're using Pyroscope Go SDK, initialize pyroscope profiler.
	_, _ = pyroscope.Start(pyroscope.Config{
		ApplicationName: "my-service",
		ServerAddress:   "http://localhost:4040",
	})

	// Your code goes here.
}
```

Now that we set up the tracer, we can create a new trace from anywhere and the profiler will automatically capture profiles for it.
```go

> ctx, span := otel.Tracer("tracerName").Start(context.Background(), "ExampleSpan")

Reviewer 2: Just if someone blindly copied the line it would use the existing context
```suggestion
ctx, span := otel.Tracer("tracerName").Start(ctx, "ExampleSpan")
```

```shell
# Make sure you also upgrade pyroscope server to version 0.14.0 or higher.
go get github.com/grafana/otel-profiling-go
```

Next you need to create and configure the tracer provider:
```go
package main

import (
	otelpyroscope "github.com/grafana/otel-profiling-go"
	"github.com/grafana/pyroscope-go"
)

func main() {
	// Initialize your tracer provider as usual.
	tp := initTracer()

	// Wrap it with otelpyroscope tracer provider.
	otel.SetTracerProvider(otelpyroscope.NewTracerProvider(tp))

	// If you're using Pyroscope Go SDK, initialize pyroscope profiler.
	_, _ = pyroscope.Start(pyroscope.Config{
		ApplicationName: "my-service",
		ServerAddress:   "http://localhost:4040",
	})

	// Your code goes here.
}
```

Now that we set up the tracer, we can create a new trace from anywhere and the profiler will automatically capture profiles for it.
```go
ctx, span := otel.Tracer("tracerName").Start(ctx, "ExampleSpan")
defer span.End()

// Your code goes here.
```

## View the span profiles in Grafana Tempo

To view the span profiles in Grafana Tempo, you need to have a Grafana instance running and a datasource configured to link Trace spans and profiles.

> Refer to the [datasource configuration documentation](https://grafana.com/docs/grafana/latest/datasources/tempo/configure-tempo-data-source/#trace-to-profiles) to see how to configure the visualization to link trace spans with profiles.

Author: ```suggestion
Refer to the [datasource configuration documentation]({{< relref "/docs/grafana/latest/datasources/tempo/configure-tempo-data-source/#trace-to-profiles" >}}) to see how to configure the visualization to link trace spans with profiles.
```

> Refer to the [datasource configuration documentation]({{< relref "/docs/grafana/latest/datasources/tempo/configure-tempo-data-source" >}}) to see how to configure the visualization to link trace spans with profiles.

Author: @knylander-grafana how do I reference this?

> Refer to the [datasource configuration documentation]({{< relref "/docs/grafana/latest/datasources/tempo/configure-tempo-data-source" >}}) to see how to configure the visualization to link trace spans with profiles.

Reviewer: I think it's by either removing latest or replacing that with <GRAFANA_VERSION>.  Reference: https://grafana.com/docs/writers-toolkit/write/links/#link-to-grafanacom-pages
```suggestion
Refer to the [datasource configuration documentation]({{< relref "/docs/grafana/<GRAFANA_VERSION>/datasources/tempo/configure-tempo-data-source" >}}) to see how to configure the visualization to link trace spans with profiles.
```

> Refer to the [datasource configuration documentation]({{< relref "/docs/grafana/<GRAFANA_VERSION>/datasources/tempo/configure-tempo-data-source" >}}) to see how to configure the visualization to link trace spans with profiles.

Reviewer: Okay. I did some testing and this seems to work: 
```suggestion
Refer to the [data source configuration documentation](/docs/grafana/datasources/tempo/configure-tempo-data-source) to see how to configure the visualization to link trace spans with profiles.
```

> Refer to the [datasource configuration documentation]({{< relref "/docs/grafana/latest/datasources/tempo/configure-tempo-data-source/#trace-to-profiles" >}}) to see how to configure the visualization to link trace spans with profiles.

Author: ```suggestion
Refer to the [datasource configuration documentation]({{< relref "/docs/grafana/latest/datasources/tempo/configure-tempo-data-source" >}}) to see how to configure the visualization to link trace spans with profiles.
```
